crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -924,47 +932,50 @@ impl<'w> EntityWorldMut<'w> {
         let archetypes = &mut self.world.archetypes;
         let storages = &mut self.world.storages;
         let components = &mut self.world.components;
-        let entities = &mut self.world.entities;
-        let removed_components = &mut self.world.removed_components;
 
-        let retained_bundle_info = self.world.bundles.init_info::<T>(components, storages);
+        let retained_bundle = self.world.bundles.init_info::<T>(components, storages);
+        // SAFETY: `retained_bundle` exists as we just initialized it.
+        let retained_bundle_info = unsafe { self.world.bundles.get_unchecked(retained_bundle) };
         let old_location = self.location;
         let old_archetype = &mut archetypes[old_location.archetype_id];
 
         let to_remove = &old_archetype
             .components()
             .filter(|c| !retained_bundle_info.components().contains(c))
             .collect::<Vec<_>>();
-        let remove_bundle_info = self
-            .world
-            .bundles
-            .init_dynamic_info(components, to_remove)
-            .0;
+        let remove_bundle = self.world.bundles.init_dynamic_info(components, to_remove);
 
-        // SAFETY: Components exist in `remove_bundle_info` because `Bundles::init_dynamic_info`
+        // SAFETY: Components exist in `remove_bundle` because `Bundles::init_dynamic_info`
         // initializes a `BundleInfo` containing all components in the to_remove Bundle.
-        unsafe {
-            Self::remove_bundle_info(
-                self.entity,
-                &mut self.location,
-                old_location,
-                remove_bundle_info,
-                archetypes,
-                storages,
-                components,
-                entities,
-                removed_components,
-            );
-        }
-
+        self.location = unsafe { self.remove_bundle(remove_bundle) };
         self
     }
 
     /// Despawns the current entity.
     pub fn despawn(self) {
-        debug!("Despawning entity {:?}", self.entity);
         let world = self.world;
-        world.flush();
+        world.flush_entities();
+        let archetype = &world.archetypes[self.location.archetype_id];
+
+        // SAFETY: Archetype cannot be mutably aliased by DeferredWorld
+        let (archetype, mut deferred_world) = unsafe {
+            let archetype: *const Archetype = archetype;
+            let world = world.as_unsafe_world_cell();
+            (&*archetype, world.into_deferred())
+        };
```

Comment:
Here is a question about architecture.

I figure this pattern could be called "laundering reference lifetime". It is used in _a lot of places_. I feel this is the result of a misaligned design.

Here in particular, you only need it to be able to call `deferred_world.trigger_on_remove` with `archetype.components()`. I feel like a better design would be a method called `trigger_archetype_on_remove(&mut World, ArchetypeId)` that both checks the `has_on_remove` and iterates over the `components()`. This way the aliasing workaroud doesn't "leak", it's part of the abstraction.

Reply:
I don't think this generalizes well as you aren't always triggering for all components in the archetype, you need some way to know what is actually removed for methods like `take` and `retain`.